pkg/server/downloader.go

```diff
@@ -59,7 +63,11 @@ func (d *BackingImageDownloader) Pull(url, filepath string, updater util.Progres
 }
 
 func (d *BackingImageDownloader) Receive(port string, filePath string, syncFileOps sparserest.SyncFileOperations) error {
-	_, cancel := context.WithCancel(context.Background())
+	if d.isDownloading() {
+		d.Cancel()
+		return fmt.Errorf("BUG: downloader gets duplicate receive requests during downloading")
+	}
+	ctx, cancel := context.WithCancel(context.Background())
 	d.Lock()
 	d.cancelFunc = cancel
 	d.Unlock()
```

Comment:
IMO, we show `check and set` value for a `shared variable` while holding the lock.

For example, assume that at the beginning we have 2 goroutines (A and B) and the `d.cancelFunc == nil`. The OS schedule the goroutines in this order:
1. A enters line 66 and check d.isDownloading(). Since `d.cancelFunc == nil`, A skips to 70
2. A enters line 66 and check d.isDownloading(). Since `d.cancelFunc == nil`, B skips to 70
3. A enters line 71 and set `d.cancelFunc = cancel`
4. A processes to download the file
5. B enters line 71 and set `d.cancelFunc` to a different  `cancel` func
6. B processes to download the file
7. Now both A and B is downloading the file. Moreover, we cannot stop B because the cancel func is belongs to B

To solve this issue, we can:
d.Lock()
        if d.cancelFunc != nil {
              return fmt.Errorf("BUG: downloader gets duplicate receive requests during downloading")
        }
	d.cancelFunc = cancel
d.Unlock()

Reply:
Oh, Sorry. My bad. Check & set should be atomic. Thank you for pointing out.